### PR TITLE
[WIP] Update Nginx role and add certbot role

### PR DESCRIPTION
**Pull request by JusteLeblanc**

### CODE REVIEW

File: generators/server/templates/devops-symfony/provisioning/roles/nginx/tasks/main.yml [hunk 1]
```diff
@@ -1,24 +1,31 @@
...
-    creates: /etc/nginx/dhparams.pem
+- name: Remove default site conf
+  file: path=/etc/nginx/sites-available/default state=absent
+  notify: reload nginx
 
-- name: restart nginx
-  service: name=nginx state=restarted
+- name: Create the Nginx configuration file - Local
+  template: src=nginx.conf.j2
+            dest=/etc/nginx/conf.d/main.conf
```
Comment (benja-M-1) [inline generators/server/templates/devops-symfony/provisioning/roles/nginx/tasks/main.yml]: En faisant ça tu surcharges toute la configuration par défaut d'nginx non ?

Par convention la configuration des applications web se met dans ```/etc/nginx/sites-available```, comme ça on peut avoir différentes configuration pour différents vhosts.

Tu peux aussi changer le nom pour ```{{ project_name }}.conf``` stp ?

File: generators/server/templates/devops-symfony/provisioning/roles/nginx/tasks/main.yml [hunk 2]
```diff
@@ -1,24 +1,31 @@
...
+            backup=yes
+  notify: reload nginx
+
+- name: Ensure that the default site is disabled
+  file: path=/etc/nginx/sites-enabled/default state=absent
+  notify: reload nginx
+
+- name: Ensure that the application site is enabled
+  file: src=/etc/nginx/conf.d/main.conf
```
Comment (benja-M-1) [inline generators/server/templates/devops-symfony/provisioning/roles/nginx/tasks/main.yml]: ```src=/etc/nginx/site-avalaible/{{ project_name }}.conf```

**Comment by foucdeg** [comment]
@JusteLeblanc je trouve dehydrated plus simple que CertBot. Exemple de provisioning en place sur Ada : https://github.com/theodo/ada-express/tree/develop/devops/provisioning/roles/dehydrated

**Comment by ghamaide** [comment]
@foucdeg @JusteLeblanc cette PR est trop cool, qui peut la tirer ?

File: generators/server/templates/devops-symfony/provisioning/roles/nginx/default/main.yml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+---
+
+nginx_server_name: "{{ inventory_hostname }}"
+nginx_ssl_dest_dir: /etc/ssl
```
Comment (MaximeThoonsen) [inline generators/server/templates/devops-symfony/provisioning/roles/nginx/default/main.yml]: not usefull anymore

File: generators/server/templates/devops-symfony/provisioning/roles/nginx/handlers/main.yml [hunk 4]
```diff
@@ -1,6 +1,4 @@
 ---
-- name: restart nginx
```
Comment (MaximeThoonsen) [inline generators/server/templates/devops-symfony/provisioning/roles/nginx/handlers/main.yml]: We don't use it in the provisioning so I suggest we remove it @JusteLeblanc

File: generators/server/templates/devops-symfony/provisioning/roles/certbot/defaults/main.yml [hunk 5]
```diff
@@ -3,4 +3,4 @@
 certbot_auto_renew: yes
 certbot_admin_email: no-reply@theodo.fr
 certbot_script: /opt/certbot-auto
-certbot_output_dir: "/etc/letsencrypt/live/{{ inventory_hostname }}"
+certbot_output_dir: "/etc/letsencrypt/live/{{ project_name }}"
```
Comment (MaximeThoonsen) [inline generators/server/templates/devops-symfony/provisioning/roles/certbot/defaults/main.yml]: I suggest to use project_name so we are more ISO through the environment + it's more simpler to understand

**Comment by JusteLeblanc** [comment]
@benja-M-1 could you check this PR and approved it if it’s ok for you? So that we can merge it ? Thanks

File: generators/server/templates/devops-symfony/provisioning/roles/certbot/defaults/main.yml (hunk 5 above)
Comment (JusteLeblanc) [reply generators/server/templates/devops-symfony/provisioning/roles/certbot/defaults/main.yml]: That's the answer 😢 
image

Comment (JusteLeblanc) [reply generators/server/templates/devops-symfony/provisioning/roles/certbot/defaults/main.yml]: @benja-M-1 
project_name = app_name que tu set au moment de la génération du projet
inventory_hostname = vagrant / staging / prod

**Comment by JusteLeblanc** [comment]
@clement-escolano this line of Nginx conf file :     
```
# Diffie-Hellman parameter for DHE ciphersuites, recommended 2048 bits
# Uncomment if you want to use your own Diffie-Hellman parameter, which can be generated with: openssl ecparam -genkey -out dhparam.pem -name prime256v1
# See https://wiki.mozilla.org/Security/Server_Side_TLS#DHE_handshake_and_dhparam
ssl_dhparam /etc/nginx/dhparams.pem;
```
is making the provisioning failed. I commented it. Should we remove it as it's not working?

**Comment by clement-escolano** [comment]
@JusteLeblanc I think this is because you removed the generation of the `dhparams` [here](https://github.com/theodo/theodo-stack-generator/pull/150/files#diff-4c6c00f26f634e83962e6feeb960c6ddL18).
You should keep the generation of the `dhparams` and this line as well 😉

**Comment by JusteLeblanc** [comment]
To avoid composer cache issues, we need to run : `chown www-data:www-data /var/www`

**Comment by JusteLeblanc** [comment]
@MaximeThoonsen I ran the generator and fixed some issues. I finally get it working. But I still have an HTTPS issue: 
image
Have you any idea on what is wrong?

**Comment by JusteLeblanc** [comment]
@benja-M-1 @MaximeThoonsen I removed https-local and jwt roles as 
- it wasn't included in this PR
- it's not working
- it could be done in another PR

I opened a specific issue for https-local : https://github.com/theodo/theodo-stack-generator/issues/191

File: doc/installation.md [hunk 6]
```diff
@@ -27,11 +27,29 @@ sudo su - www-data
...
+
+- Update DATABASE_URL in `.env` file with `db_user`, `db_password` and `db_name`
+- If you are using PostgresSQL, you need:
+     - to replace in .env file `mysql` by `pgsql` and `3306`by `5432`
+     - to update `doctrine.yaml` these way :
+
+```
+    dbal:
+         # configure these for your database server
+         driver: 'pdo_mysql'
```
Comment (benja-M-1) [inline doc/installation.md]: ```pdo_pgsql``` no?

File: doc/installation.md [hunk 7]
```diff
@@ -27,11 +27,29 @@ sudo su - www-data
...
+TRUSTED_HOSTS='{myAppName}.local, localhost, 10.0.0.10'
   ```
 
 - Add API Platform if you need it `cd api && composer req api`
 - Install and configure the php code sniffer with SymfonyCustom coding standard `composer require --dev vincentlanglet/symfony3-custom-coding-standard && vendor/bin/phpcs --config-set installed_paths ../../vincentlanglet/symfony3-custom-coding-standard`
+
+- Update DATABASE_URL in `.env` file with `db_user`, `db_password` and `db_name`
+- If you are using PostgresSQL, you need:
+     - to replace in .env file `mysql` by `pgsql` and `3306`by `5432`
+     - to update `doctrine.yaml` these way :
```
Comment (benja-M-1) [inline doc/installation.md]: these --> this

File: doc/installation.md [hunk 8]
```diff
@@ -27,11 +27,29 @@ sudo su - www-data
...
+```
+    dbal:
+         # configure these for your database server
+         driver: 'pdo_mysql'
+         server_version: '5.7'
+         charset: utf8
+         default_table_options:
+             charset: utf8
+             collate: utf8_unicode_ci
+         url: '%env(resolve:DATABASE_URL)%'
```
Comment (benja-M-1) [inline doc/installation.md]: from the [.env of Symfony Flex](https://github.com/symfony/recipes/blob/master/doctrine/doctrine-bundle/1.6/config/packages/doctrine.yaml), you need to change the charset configuration and the collation. I think this need some explainations in the doc, don't you think?

File: generators/server/templates/devops-symfony/provisioning/roles/nginx/default/main.yml [hunk 9]
```diff
@@ -0,0 +1,6 @@
...
+nginx_strong_dh_group: yes  # Strongly recomended in production. See weakdh.org.
+nginx_use_letsencrypt: yes
```
Comment (benja-M-1) [inline generators/server/templates/devops-symfony/provisioning/roles/nginx/default/main.yml]: is didn't find where those parameters are used, I think they are useless

**Comment by MaximeThoonsen** [comment]
@JusteLeblanc Ouai ça marche sur mon ubuntu mais pas sur les MAC. Il bloque les certificats self-signed

**Comment by foucdeg** [comment]
@MaximeThoonsen [mkcert](https://github.com/FiloSottile/mkcert) marche nickel

**Comment by JusteLeblanc** [comment]
We can't test certbot role on AWS EC2 instances as LetsEncrypt blocks certificate creations for EC2 domains: https://community.letsencrypt.org/t/policy-forbids-issuing-for-name-on-amazon-ec2-domain/12692. 

The solution @MaximeThoonsen suggested is to have a fixed IP address on AWS, to create a DNS record to a `theo.do` domain name. It means we need to associate this fix IP each time we create the EC2 instance in test flow. @MaximeThoonsen how could we automate this task?

cc @benja-M-1 @tcheymol

**Comment by JusteLeblanc** [comment]
> We can't test certbot role on AWS EC2 instances as LetsEncrypt blocks certificate creations for EC2 domains: https://community.letsencrypt.org/t/policy-forbids-issuing-for-name-on-amazon-ec2-domain/12692.
> 
> The solution @MaximeThoonsen suggested is to have a fixed IP address on AWS, to create a DNS record to a `theo.do` domain name. It means we need to associate this fix IP each time we create the EC2 instance in test flow. @MaximeThoonsen how could we automate this task?
> 
> cc @benja-M-1 @tcheymol

@MaximeThoonsen We need to add it in test script using Ansible or AWS API.

**Pull request merged**